In [6]:
from datetime import datetime
from datetime import timedelta

feb_8_1986: datetime = datetime(1986, 2, 8)
jul_28_2061: datetime = datetime(2061, 7, 28)

num_days_between: int = (jul_28_2061 - feb_8_1986).days
print(num_days_between)

for i in range(1, 6):
    next_year: datetime = feb_8_1986 + timedelta(days=num_days_between * i)
    print(f"Halley's comet will next appear {next_year.strftime("%Y-%m-%d")}")

for i in range(1, 21):
    last_year: datetime = feb_8_1986 - timedelta(days=num_days_between * i)
    print(f"Halley's comet last appeared {last_year.strftime("%Y-%m-%d")}")

27564
Halley's comet will next appear 2061-07-28
Halley's comet will next appear 2137-01-15
Halley's comet will next appear 2212-07-05
Halley's comet will next appear 2287-12-23
Halley's comet will next appear 2363-06-12
Halley's comet last appeared 1910-08-22
Halley's comet last appeared 1835-03-04
Halley's comet last appeared 1759-09-14
Halley's comet last appeared 1684-03-26
Halley's comet last appeared 1608-10-07
Halley's comet last appeared 1533-04-20
Halley's comet last appeared 1457-10-31
Halley's comet last appeared 1382-05-13
Halley's comet last appeared 1306-11-24
Halley's comet last appeared 1231-06-06
Halley's comet last appeared 1155-12-18
Halley's comet last appeared 1080-06-29
Halley's comet last appeared 1005-01-10
Halley's comet last appeared 929-07-23
Halley's comet last appeared 854-02-02
Halley's comet last appeared 778-08-16
Halley's comet last appeared 703-02-27
Halley's comet last appeared 627-09-09
Halley's comet last appeared 552-03-21
Halley's comet last appea

# Negative Dates

The Python datetime module crashes when it tries to work with dates less than 1 CE. To get around this we need external modules to handle computing dates. 

There are several, with varying degrees of intelligence about calendars and leap years. For this example we will use NumPy's datetime64 module. 

In [ ]:
import numpy as np

for i in range(1, 31):
    np_date = np.datetime64(feb_8_1986) - np.timedelta64(num_days_between * i, 'D')
    print(f"Halley's comet last appeared {np.datetime_as_string(np_date, unit='D')}")

Halley's comet last appeared 1910-08-22
Halley's comet last appeared 1835-03-04
Halley's comet last appeared 1759-09-14
Halley's comet last appeared 1684-03-26
Halley's comet last appeared 1608-10-07
Halley's comet last appeared 1533-04-20
Halley's comet last appeared 1457-10-31
Halley's comet last appeared 1382-05-13
Halley's comet last appeared 1306-11-24
Halley's comet last appeared 1231-06-06
Halley's comet last appeared 1155-12-18
Halley's comet last appeared 1080-06-29
Halley's comet last appeared 1005-01-10
Halley's comet last appeared 0929-07-23
Halley's comet last appeared 0854-02-02
Halley's comet last appeared 0778-08-16
Halley's comet last appeared 0703-02-27
Halley's comet last appeared 0627-09-09
Halley's comet last appeared 0552-03-21
Halley's comet last appeared 0476-10-01
Halley's comet last appeared 0401-04-14
Halley's comet last appeared 0325-10-26
Halley's comet last appeared 0250-05-08
Halley's comet last appeared 0174-11-18
Halley's comet last appeared 0099-05-31


Compare your calculations against a table of the known dates:

| Perihelion |
|---|
| 9 February 1986 |
| 20 April 1910 |
| 16 November 1835 |
| 13 March 1759 |
| 15 September 1682 |
| 27 October 1607 |
| 26 August 1531 |
| 9 June 1456 |
| 10 November 1378 |
| 25 October 1301 |
| 28 September 1222 |
| 18 April 1145 |
| 20 March 1066 |
| 5 September 989 |
| 18 July 912 |
| 28 February 837 |
| 20 May 760 |
| 2 October 684 |
| 15 March 607 |
| 27 September 530 |
| 28 June 451 |
| 16 February 374 |
| 20 April 295 |
| 17 May 218 |
| 22 March 141 |
| 25 January AD 66 |
| 10 October 12 BC |
| 6 August 87 BC |
| 12 November 164 BC |
| 25 May 240 BC |